### PR TITLE
read write cycle testing

### LINKED ISSUE #19: tests don't delete `temp` folder before running
this means that tests can pass or fail depending on the contents of the `temp` folder. not good!

### PR DESCRIPTION
**Pull request by ix0rai**
hopefully the last pr of the 1.2 cycle. adds extensive testing for read/write cycle to ensure no data is lost during reading or writing
fixes #19 

contains one API addition: `MetadataContainer#metadata()`
```java

         /**
	 * @return a map of all metadata attached to this node
	 */
	Map<MetadataType<?, ?>, Object> metadata();
```

TODO:
- [x] add tests that simply read the file and compare against a string, to ensure formatting is proper - decided against doing this
- [x] expand read/write tests

notes:
- diff is huge because it's 2024 and licenses need updating
- shouldn't change functionality of anything. I haven't found any issues in writing these tests (so far)

### CODE REVIEW

**Review by OroArmor — APPROVED** [review]

File: src/main/java/org/quiltmc/config/impl/util/ConfigsImpl.java [hunk 1]
```diff
@@ -40,6 +39,13 @@ public static void put(String familyId, Config config) {
 		CONFIGS.computeIfAbsent(familyId, id -> new TreeMap<>()).put(config.id(), config);
 	}
 
+	/**
+	 * Not intended to be used for anything other than testing.
```
Comment (IotaBread) [inline src/main/java/org/quiltmc/config/impl/util/ConfigsImpl.java:43]: `@org.jetbrains.annotations.TestOnly` is exactly for this

File: src/test/java/org/quiltmc/config/reflective/TestEscapingConfig.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import org.quiltmc.config.api.values.ValueList;
+import org.quiltmc.config.api.values.ValueMap;
+
+public class TestEscapingConfig extends ReflectiveConfig {
+	@SerializedName("gaming.awesome")
+	public final TrackedValue<String> awesome = this.value("gaming@1");
+
+	@SerializedName("servers.server.served")
+	public final TrackedValue<ValueMap<String>> servers = this.value(
+		ValueMap.builder("").put("s@dir/otherdir/thirddir", "rai minecraft").put("m@www.raiminecraft.com", "minecraft").build());
```
Comment (IotaBread) [inline src/test/java/org/quiltmc/config/reflective/TestEscapingConfig.java:15]: Love these

**Review by IotaBread — APPROVED** [review]

File: src/test/java/org/quiltmc/config/reflective/TestEscapingConfig.java (hunk 2 above)
Comment (ix0rai) [reply src/test/java/org/quiltmc/config/reflective/TestEscapingConfig.java:31]: I have to make my mark on humanity somehow

File: src/main/java/org/quiltmc/config/impl/util/ConfigsImpl.java (hunk 1 above)
Comment (ix0rai) [reply src/main/java/org/quiltmc/config/impl/util/ConfigsImpl.java]: oh neat!

**Pull request merged**